sdlib/d/gc/slab.d

```diff
@@ -23,6 +25,159 @@ struct SlabAllocGeometry {
 	}
 }
 
+struct SlabAllocInfo {
+public:
+	bool hasMetaData = false;
+
+	this(PageDescriptor pd, const void* ptr) {
+		assert(pd.isSlab(), "Expected a slab!");
+
+		e = pd.extent;
+		sg = SlabAllocGeometry(pd, ptr);
+		allowsMetaData = sizeClassSupportsMetadata(pd.sizeClass);
+		hasMetaData = allowsMetaData && e.hasMetaData(sg.index);
+	}
+
+	@property
+	auto address() {
+		return sg.address;
+	}
+
+	@property
+	size_t slotCapacity() {
+		return sg.size;
+	}
+
+	@property
+	size_t usedCapacity() {
+		return allowsMetaData ? sg.size - freeSpace : 0;
```

Comment:
Isn't 0 here a bug?

Reply:
Inf act, free space checks for the presence of metadata, which is gated by allowsMetaData already, so this whole ternary is useless, no?

Reply:
I don't think there is any significant change in behavior here. Especially since a used capacity of zero is something we might want to be valid in the future.